jxl/src/api/output.rs

```diff
@@ -5,95 +5,169 @@
-    ) -> impl DerefMut<Target = [&'a mut [MaybeUninit<u8>]]> {
-        assert!(self.len() >= shape.0 * shape.1 * bytes_per_pixel);
-        let mut ret = vec![];
-        let mut slc = self;
-        for _ in 0..shape.1 {
-            let (part, new_slc) = slc.split_at_mut(shape.0 * bytes_per_pixel);
-            ret.push(part);
-            slc = new_slc;
+    /// Creates a new JxlOutputBuffer from raw pointers.
+    /// It is guaranteed that `buf` will never be used to write uninitalized data.
+    ///
+    /// # Safety
+    /// - `buf` must be valid for writes for all bytes in the range
+    ///   `buf[i*bytes_between_rows..i*bytes_between_rows+bytes_per_row]` for all values of `i`
+    ///   from `0` to `num_rows-1`.
+    /// - The bytes in these ranges must not be accessed as long as the returned `Self` is in scope.
+    /// - All the bytes in those ranges (and in between) must be part of the same allocated object.
+    pub unsafe fn new_from_ptr(
+        buf: *mut MaybeUninit<u8>,
+        num_rows: usize,
+        bytes_per_row: usize,
+        bytes_between_rows: usize,
+    ) -> Self {
+        Self::check_vals(num_rows, bytes_per_row, bytes_between_rows);
+        // Safety: caller guarantees the buf-related invariants, and other safety invariants are
+        // checked in check_vals.
+        Self {
+            buf,
+            bytes_per_row,
+            bytes_between_rows,
+            num_rows,
+            _ph: PhantomData,
         }
-        ret
     }
-}
 
-impl<'a> JxlOutputBuffer<'a> for [&'a mut [u8]] {
-    unsafe fn get_row_buffers(
-        &'a mut self,
-        shape: (usize, usize),
-        bytes_per_pixel: usize,
-    ) -> impl DerefMut<Target = [&'a mut [MaybeUninit<u8>]]> {
-        assert_eq!(self.len(), shape.1, "Incorrect number of rows provided");
-        for row in self.iter() {
-            assert!(
-                row.len() >= shape.0 * bytes_per_pixel,
-                "A row buffer is too short"
-            );
+    pub fn new(buf: &'a mut [u8], num_rows: usize, bytes_per_row: usize) -> Self {
+        Self::new_uninit(
+            // Safety: `new_uninit` guarantees that no uninit data is ever written to the passed-in
+            // slice. Moreover, `T` and `MaybeUninit<T>` have the same memory layout.
+            unsafe { std::mem::transmute::<&'a mut [u8], &'a mut [MaybeUninit<u8>]>(buf) },
+            num_rows,
+            bytes_per_row,
+        )
+    }
+
+    /// Creates a new JxlOutputBuffer from a slice of uninit data.
+    /// It is guaranteed that `buf` will never be used to write uninitalized data.
+    pub fn new_uninit(
+        buf: &'a mut [MaybeUninit<u8>],
+        num_rows: usize,
+        bytes_per_row: usize,
+    ) -> Self {
+        Self::check_vals(num_rows, bytes_per_row, bytes_per_row);
+        assert!(buf.len() >= bytes_per_row * num_rows);
+        // Safety: buf-related invariants follow from this function mutably borrowing a slice, and
```

Comment:
shouldn't here be an unsafe block?

Reply:
Unfortunately [unsafe fields](https://github.com/rust-lang/rfcs/pull/3458) are not stable yet, so upholding satisfying safety invariants doesn't need unsafe blocks :(